### ISSUE TITLE
Bug with optimizer_v2.OptimizerV2.set_weights method 

### ISSUE THREAD

**Issue by suneeta-mall** [tensorflow/tensorflow#50859]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
 None
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Linux (Ubuntu/Debina)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- NA
- TensorFlow installed from (source or binary):
Binary
- TensorFlow version (use command below):
- Python version: 2.4.1 & 2.4.x
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
11.0
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`
2.4.1

**Describe the current behavior**
I am seeing an issue in setting Adam optimizer weights in Optimizer v2 implementation. Looking at the code here https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/optimizer_v2/adam.py#L154-L160, it seems to be able to set the weights the current optimizer must have been initialized and set right size in terms of array length. If this assertion is correct setting weights `on_train_begin` or prior to first gradinet calculation is a bit too soon to set optimizers weights. 
In earlier versions, this was not a problem and optimizer set_weights use to work at `on_train_begin`.  
I think this a bug in how slots and Mi is structured

Code to reproduce the issue is here: 
https://gist.github.com/suneeta-mall/b6086a6b252075f48936a5f80f040b77
To run try:
```
python debug_optimizer_issue.py
```
& then try again:
```
python debug_optimizer_issue.py --start_epoch 2
```

The first run finishes and second one, where its suppose to start from epoch 2,  fails with error as:
```
ValueError: You called `set_weights(weights)` on optimizer Adam with a  weight list of length 9, but the optimizer was expecting 0 weights. Provided weights: [120, array([[ 0.0000000e+00,  0.0000000e+00,  0.0...
```

I looked around in Adam V2 code to see if there are any other APIs that can help with creating slots etc but I didn't see any that does not require immplementation details. 

As a crazy thought, I use array extension on actual weights to bypass the bug in set_weights like this:
```
model.optimizer.weights.extend(_ref_model.optimizer.get_weights())
```
This lets past the training regime, but the next model save errors into:
```
AttributeError: 'numpy.int64' object has no attribute 'name'
```

Worth noting, same problem exists with SGD, thats met with following related error:
```
ValueError: You called `set_weights(weights)` on optimizer SGD with a  weight list of length 1, but the optimizer was expecting 0 weights. Provided weights: [120]...
```

**Describe the expected behavior**
I expect the set_weights to work naturally. 
**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by UsharaniPagadala** [comment]
@suneeta-mall 

Could you please share the colab gist with all the dependencies to analyse the issue better.Thanks

**Comment by suneeta-mall** [comment]
@UsharaniPagadala  https://colab.research.google.com/drive/1x03x1trJQcOXG07otYnLyyFjCJSxjUG-?usp=sharing

**Comment by UsharaniPagadala** [comment]
@suneeta-mall 

Could you please refer the similar issues [link](https://github.com/keras-team/keras/issues/7229 ) and also refer this [optimizer_v2.py Line 989](https://github.com/tensorflow/tensorflow/blob/bd6b557c02a5cc1d094a7bb180b9779121a58520/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L989) and let us know if it helps.Thanks

**Comment by suneeta-mall** [comment]
Hi @UsharaniPagadala,
Thanks for looking into this issue. IMU, reference of https://github.com/keras-team/keras/issues/7229 is not directly related here as I don't have a problem in parameter weight setting. 

The problem is specific to optimizers' weights. I am aware of the length check code in V2 Optimizer here https://github.com/tensorflow/tensorflow/blob/bd6b557c02a5cc1d094a7bb180b9779121a58520/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L988 that you have linked. I am also aware that in V2 the initialization does not happen until the first step is made. 

However, I am unclear what you are suggesting. Are you saying this is by design or suggesting a solution? 
### This cant be by design
To be able to cleanly restart a training, optimizers' weights need to be loaded before the first step and not after. So, IMU, the optimizers.set_weights() call should be able to handle the loading of weights beyond current weight length & source weight length equality restriction.  Whether that's done implicitly in set_weights call or another alternative simple initialization API is provided so that caller can initialize optimizer.param capacity to hold incoming weights. Otherwise, there is no way to restart training using optimizer weights.
Stressing to the point that this use to work as expected in V1 optimizers. 

### Possible workaround 
My apologies if I missed but I am not seeing a workaround suggested here. Like I said earlier, setting after a step has been made is a tad bit too late to set optimizers. I also don't see callbacks that may be of use in this case but did not find any.

**Comment by UsharaniPagadala** [comment]
@suneeta-mall 

I was able to reproduce the code shared in tf2.5 with no errors.Please find the [gist](https://colab.research.google.com/gist/UsharaniPagadala/45e8e99b7870e1f81b115d88c7f72ed3/tf-bug-50859-v2optimizer.ipynb) and let us know if it helps.Thanks

**Comment by suneeta-mall** [comment]
@UsharaniPagadala That's excellent .. thanks for testing it for now .. any chance we backport the fix to 2.4?

**Comment by UsharaniPagadala** [comment]
@suneeta-mall 

We don't frequently backport the fix to the previous versions. Recommend you to upgrade to the latest stable Version.Thanks

**Comment by UsharaniPagadala** [comment]
@suneeta-mall 
Could you please move this to closed status as it is resolved.Thanks

**Comment by suneeta-mall** [comment]
@UsharaniPagadala I disagree that its fixed in 2.5. I tried and I saw the same issue on 2.5 then had a look at your notebook and can see you missed running the actual step tat restarts the training .. 
see here https://colab.research.google.com/gist/UsharaniPagadala/45e8e99b7870e1f81b115d88c7f72ed3/tf-bug-50859-v2optimizer.ipynb#scrollTo=YXqWUOIpg44F

**Comment by suneeta-mall** [comment]
@Saduf2019 Are you talking about `load_model` with compile=True? I am not sure if thats a solution as the author itself says it works and it also does not? Nevertheless, my usecase is slightly different so I need a way to set optimizer weights. 

Do you not agree that this is a bug in V2 Optimizer that needs fixing?

**Comment by irasit** [comment]
Seems the issue in keras-team/keras repo https://github.com/keras-team/keras/issues/15298 is closed without fix. Is there a plan to fix this behavior?
